ddtrace/span.py

```diff
@@ -478,7 +478,7 @@ def set_exc_info(self, exc_type, exc_val, exc_tb):
         # readable version of type (e.g. exceptions.ZeroDivisionError)
         exc_type_str = "%s.%s" % (exc_type.__module__, exc_type.__name__)
 
-        self.meta[ERROR_MSG] = str(exc_val)
+        self.meta[ERROR_MSG] = str(exc_val) if six.PY2 and isinstance(exc_val.message, str) else stringify(exc_val)
```

Comment:
why doesn't `stringify(exc_val)` work?

`compat.stringify` is just `six.text_type`


```
Python 3.9.9 (main, Nov 21 2021, 03:23:42)
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> e = Exception("DataDog/水")
>>> import six
>>> six.text_type(e)
'DataDog/水'
```

```
Python 2.7.16 (default, Sep  6 2021, 07:39:44)
[GCC Apple LLVM 12.0.5 (clang-1205.0.19.59.6) [+internal-os, ptrauth-isa=deploy on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> e = Exception(u"DataDog/水")
>>> import six
>>> six.text_type(e)
u'DataDog/\u6c34'
```

You likely just need to adjust the assertion to match what Python2 is expecting?

Reply:
In python2 unicode(Exception("DataDog/水")) fails and throws the following error: `UnicodeDecodeError: 'ascii' codec can't decode byte 0xe6 in position 8: ordinal not in range(128)`. However unicode(Exception(u"DataDog/水")) runs without any issues. Since stringify == unicode for python2 it would also throw this error. 

I'll add a test case for bytestrings for python2 to see if we run into the same error.

Instead of using this approach I considered setting the exception value to Exception.message in python2 (exc_val =  exc_val.message if six.PY2) but I ran into another issue. Some subclasses of Exception do not set the message field (ex. [werkzeug.exceptions.InternalServerError](https://github.com/pallets/werkzeug/blob/main/src/werkzeug/exceptions.py#L78) sets response and description fields but sets message to an empty string. This caused some tests which test the ERROR_MSG meta tag the to fail.

Reply:
Although, this doesn't work when the Python encoding is set to `utf-8` for Python 2

```
$ PYTHONIOENCODING=utf8 python2
>>> type("DataDog/水")
<type 'str'>
>>> import six
>>> six.text_type(Exception("DataDog/水"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe6 in position 8: ordinal not in range(128)
>>> str(Exception("DataDog/水"))
'DataDog/\xe6\xb0\xb4'

Reply:
Ya that's why I added the PY2 check. I spent most of today experiencing different errors with decoding text types. 

In python2 unicode(Exception("DataDog/水")) fails and throws the following error: UnicodeDecodeError: 'ascii' codec can't decode byte 0xe6 in position 8: ordinal not in range(128). However unicode(Exception(u"DataDog/水")) runs without any issues. Since stringify == unicode for python2 it would also throw this error.
I'll add a test case for when the exception message is a bytestring to see if we run into a similar error in python2.
Instead of using this approach I considered setting the exception value to Exception.message in python2 (exc_val = exc_val.message if six.PY2) but I ran into another issue. Some subclasses of Exception do not set the message field (ex. werkzeug.exceptions.InternalServerError sets response and description fields but sets message to an empty string. We could simplify this code to self.meta[ERROR_MSG] = exc_val.message if six.PY2 else stringify(exc_val) but this would cause some spans to lose their error message

Reply:
`unicode(Exception)` is not what you want IMHO. `Exception` are meant to implement a `__str__` but not `__unicode__` so you'll always fallback to `__str__` and get bytes.

If `meta` value should be:

1. Unicode, then you need to call `compat.to_unicode()`
2. Raw bytes, then you need `str(Exception())` in Python 2 and call `str(Exception()).encode("utf-8")` in Python 3